lib/src/booleans/boolean_overlap.dart

```diff
@@ -0,0 +1,213 @@
+import 'package:geojson_equality/geojson_equality.dart';
+
+import '../../helpers.dart';
+import '../../line_segment.dart';
+import '../invariant.dart';
+import '../line_intersect.dart';
+import '../line_overlap.dart';
+
+/**
+ * Compares two geometries of the same dimension and returns true if their intersection set results in a geometry
+ * different from both but of the same dimension. It applies to Polygon/Polygon, LineString/LineString,
+ * Multipoint/Multipoint, MultiLineString/MultiLineString and MultiPolygon/MultiPolygon.
+ *
+ * In other words, it returns true if the two geometries overlap, provided that neither completely contains the other.
+ *
+ * @name booleanOverlap
+ * @param  {Geometry|Feature<LineString|MultiLineString|Polygon|MultiPolygon>} feature1 input
+ * @param  {Geometry|Feature<LineString|MultiLineString|Polygon|MultiPolygon>} feature2 input
+ * @returns {boolean} true/false
+ * @example
+ * var poly1 = turf.polygon([[[0,0],[0,5],[5,5],[5,0],[0,0]]]);
+ * var poly2 = turf.polygon([[[1,1],[1,6],[6,6],[6,1],[1,1]]]);
+ * var poly3 = turf.polygon([[[10,10],[10,15],[15,15],[15,10],[10,10]]]);
+ *
+ * turf.booleanOverlap(poly1, poly2)
+ * //=true
+ * turf.booleanOverlap(poly2, poly3)
+ * //=false
+ */
+booleanOverlap(GeometryObject feature1, GeometryObject feature2) {
+  var geom1 = getGeom(feature1);
+  var geom2 = getGeom(feature2);
+  var type1 = geom1.type;
+  var type2 = geom2.type;
+
+  if ((type1 == MultiPoint && type2 != MultiPoint) ||
+      ((type1 == LineString || type1 == MultiLineString) &&
+          type2 != LineString &&
+          type2 != MultiLineString) ||
+      ((type1 == Polygon || type1 == MultiPolygon) &&
+          type2 != Polygon &&
+          type2 != MultiPolygon)) {
+    throw Exception("features must be of the same type");
+  }
+  if (type1 == Point) throw Exception("Point geometry not supported");
+
+  // features must be not equal
+  var equality = Equality(precision: 6);
+  if (equality.compare(feature1, feature2)) return false;
+
+  var overlap = 0;
+
+  switch (type1) {
+    case MultiPoint:
+      for (var i = 0; i < (geom1 as MultiPoint).coordinates.length; i++) {
+        for (var j = 0; j < (geom2 as MultiPoint).coordinates.length; j++) {
+          var coord1 = geom1.coordinates[i];
+          var coord2 = geom2.coordinates[j];
+          if (coord1[0] == coord2[0] && coord1[1] == coord2[1]) {
+            return true;
+          }
+        }
+      }
```

Comment:
we can get rid of this typecheck at the beginning if we later in the function replace the switch on the enum with `if(type is MultiPoint)` e.g. We **need** to do it consistently throughout the lib.